File: apps/app-frontend/src/helpers/pack.js
/**
 * All theseus API calls return serialized values (both return values and errors);
 * So, for example, addDefaultInstance creates a blank Profile object, where the Rust struct is serialized,
 *  and deserialized into a usable JS object.
 */
import { invoke } from '@tauri-apps/api/core'

import { create } from './profile'

// Installs pack from a version ID
export async function create_profile_and_install(
	projectId,
	versionId,
	packTitle,
	iconUrl,
	createInstanceCallback = () => {},
) {
	const location = {
		type: 'fromVersionId',
		project_id: projectId,
		version_id: versionId,
		title: packTitle,
		icon_url: iconUrl,
	}
	const profile_creator = await invoke('plugin:pack|pack_get_profile_from_pack', { location })
	const profile = await create(
		profile_creator.name,
		profile_creator.gameVersion,
		profile_creator.modloader,
		profile_creator.loaderVersion,
		null,
		true,
	)
	createInstanceCallback(profile)

	return await invoke('plugin:pack|pack_install', { location, profile })
}

export async function install_to_existing_profile(projectId, versionId, title, profilePath) {
	const location = {
		type: 'fromVersionId',
		project_id: projectId,
		version_id: versionId,
		title,
	}
	return await invoke('plugin:pack|pack_install', { location, profile: profilePath })
}

// Installs pack from a path
export async function create_profile_and_install_from_file(path) {
	const location = {
		type: 'fromFile',
		path: path,
	}
	const profile_creator = await invoke('plugin:pack|pack_get_profile_from_pack', { location })
	const profile = await create(
		profile_creator.name,
		profile_creator.gameVersion,
		profile_creator.modloader,
		profile_creator.loaderVersion,
		null,
		true,
	)
	return await invoke('plugin:pack|pack_install', { location, profile })
}

export async function get_catalog() {
	return await invoke('plugin:pack|pack_get_catalog')
}

export async function install_catalog_pack(packId) {
	return await invoke('plugin:pack|pack_install_catalog', { packId })
}

export async function cancel_catalog_install(jobId) {
	return await invoke('plugin:pack|pack_cancel_catalog_install', { jobId })
}

export async function get_catalog_install_status(jobId) {
	return await invoke('plugin:pack|pack_catalog_install_status', { jobId })
}

export async function verify_catalog_install(profilePath) {
	return await invoke('plugin:pack|pack_verify_catalog_install', { profilePath })
}

File: apps/app/src/api/pack.rs
use crate::api::Result;

use theseus::{
    pack::{
        install_from::{CreatePackLocation, CreatePackProfile},
        install_mrpack::install_zipped_mrpack,
    },
    prelude::*,
};

pub fn init<R: tauri::Runtime>() -> tauri::plugin::TauriPlugin<R> {
    tauri::plugin::Builder::new("pack")
        .invoke_handler(tauri::generate_handler![
            pack_install,
            pack_get_profile_from_pack,
            pack_get_catalog,
            pack_install_catalog,
            pack_cancel_catalog_install,
            pack_catalog_install_status,
            pack_verify_catalog_install,
        ])
        .build()
}

#[tauri::command]
pub async fn pack_get_catalog() -> Result<pack::catalog::CatalogResponse> {
    Ok(pack::catalog::get_catalog().await?)
}

#[tauri::command]
pub async fn pack_install_catalog(pack_id: &str) -> Result<String> {
    Ok(pack::catalog_install::start_catalog_install(pack_id).await?)
}

#[tauri::command]
pub fn pack_cancel_catalog_install(job_id: &str) -> Result<()> {
    Ok(pack::catalog_install::cancel_catalog_install(job_id)?)
}

#[tauri::command]
pub fn pack_catalog_install_status(
    job_id: &str,
) -> Result<pack::catalog_install::CatalogInstallStatus> {
    Ok(pack::catalog_install::catalog_install_status(job_id)?)
}

#[tauri::command]
pub async fn pack_verify_catalog_install(profile_path: &str) -> Result<bool> {
    Ok(
        pack::catalog_install::verify_installed_catalog_pack(profile_path)
            .await?,
    )
}

#[tauri::command]
pub async fn pack_install(
    location: CreatePackLocation,
    profile: String,
) -> Result<String> {
    Ok(install_zipped_mrpack(location, profile).await?)
}

#[tauri::command]
pub fn pack_get_profile_from_pack(
    location: CreatePackLocation,
) -> Result<CreatePackProfile> {
    Ok(pack::install_from::get_profile_from_pack(location))
}
